### PR TITLE
DrivAerML: volume training ablation — surface+volume vs surface-only

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Our current DrivAerML recipe trains on **surface points only** (`--surface-only-drivaerml`, 50k surface points per case). AB-UPT — which achieves our 3.71% target — trains on **both surface AND volume points**. Adding volume flow-field context may help the model predict surface quantities better by giving it information about the surrounding flow structure (boundary layer thickness, wake topology, separation regions) that isn't directly visible from surface data alone.

The human research team (issue #3020) specifically directs: **"Matched 16k/16k surface-volume only if the joint path is stable."**

This experiment is the first test of volume training on DrivAerML in the radford programme.

### Runs (4 DM runs — surface+volume ablation)

Use `--wandb_group dm_volume_ablation` for all runs. All use the golden DM config (AdamW lr=5e-4, T_max=30, no-EMA, Fourier, 4L/512d).

**Run 1 — 50k surface + 16k volume (add volume to baseline):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --no-surface-only-drivaerml \
  --drivaerml-train-surface-points 50000 --drivaerml-train-volume-points 16000 \
  --drivaerml-eval-surface-points 50000 --drivaerml-eval-volume-points 16000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm_volume_ablation
```
Adds 16k volume points on top of the standard 50k surface budget. Total points per case: 66k. Tests whether volume context helps surface prediction at a manageable throughput cost.

**Run 2 — 16k surface + 16k volume (matched budget per human directive):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --no-surface-only-drivaerml \
  --drivaerml-train-surface-points 16000 --drivaerml-train-volume-points 16000 \
  --drivaerml-eval-surface-points 50000 --drivaerml-eval-volume-points 16000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm_volume_ablation
```
The human-specified matched 16k/16k budget. Total 32k points per case (vs baseline 50k surface-only). Faster epochs → more epochs per 360-min → potentially deeper convergence despite lower per-epoch resolution.

**Run 3 — 16k surface only (throughput control):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 16000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm_volume_ablation
```
Same 16k surface as Run 2 but NO volume points. Isolates the volume contribution from the throughput effect. If Run 2 > Run 3, volume points are helping. If Run 3 > Run 2, the volume points are noise.

**Run 4 — 50k surface only (baseline control, FULL EVAL):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --wandb_group dm_volume_ablation
```
**NOTE: NO --max-eval-batches.** This is the surface-only baseline with full-coverage eval for paper-facing comparison. Same as the current champion config (#2898).

### GPU Priority
- **GPUs 0-3:** Run all 4 DM runs in parallel (batch_size=1, 1 GPU each)
- Run 4 (full-eval baseline control) should complete without interruption

### Important Notes
- If `--no-surface-only-drivaerml` causes errors or crashes, report immediately — this flag may require additional data preparation for volume mesh points.
- **Smoke test first:** Run each config for 3 epochs before committing to full 360-min budget. If any crash, stop and report.
- Use full-coverage eval (no --max-eval-batches) on Run 4 only. Runs 1-3 use --max-eval-batches 200 for iteration speed.

### Reporting
For each run, report:
1. Best `val_primary/surface_rel_l2_pct` and the epoch
2. Final (last-epoch) val metric
3. `test_primary/surface_rel_l2_pct` at final epoch
4. Epochs completed in the 360-min budget (throughput comparison is critical)
5. Whether volume runs crashed/diverged

### Key Comparisons
- **Run 1 vs Run 4:** Does adding volume context to the full 50k surface budget help?
- **Run 2 vs Run 3:** Does volume context help at the reduced 16k surface budget?
- **Run 3 vs Run 4:** What's the throughput/accuracy tradeoff of 16k vs 50k surface-only?

## Baseline

**DrivAerML (target to beat):**
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B run: bht6h42t (#2898 piccolo)
- Config: AdamW lr=5e-4, T_max=30, no-EMA, Fourier, 4L/512d, surface-only 50k
- External target: **3.71%** (AB-UPT trains on surface+volume)
- Val/test gap: val=3.997% but test≈5.93%

**Note on AB-UPT:** The 3.71% target was achieved by AB-UPT which uses geometry-separated encoding and trains on both surface AND volume data. Our surface-only approach might be missing the volume signal that AB-UPT captures. This experiment tests that hypothesis directly.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Run Summary Table

| Run | Config | Epochs | Best val_primary/surface_rel_l2_pct | Epoch | Best test_primary/surface_rel_l2_pct | Diverged? | W&B run | Peak VRAM |
|-----|--------|--------|-------------------------------------|-------|--------------------------------------|-----------|---------|-----------|
| 1 | 50k surf + 16k vol | 999 (epoch cap) | 58.55% | 169 | 59.27% | **YES** (NaN) | eurdv6ot | ~7.4 GB |
| 2 | 16k surf + 16k vol | 999 (epoch cap) | 64.86% | 81 | 62.61% | **YES** (NaN) | pbsrazga | ~4.1 GB |
| 3 | 16k surf only | 484 (360-min timeout) | 14.31% | 130 | 15.28% | No (grad spike) | ytrv9t7y | ~3.5 GB |
| 4 | 50k surf only (full eval) | 46 (360-min timeout) | 13.48% | 45 | 12.94% | No | pvjquyxe | ~6.0 GB |

**Baseline (PR #2898):** val 3.997% @ ep467, test ~5.93%, W&B bht6h42t

### Key Comparisons

**Run 1 vs Run 4 — Does adding volume context to 50k surface help?**
NO. Catastrophically worse. 58.55% vs 13.48%. Adding 16k volume points on top of 50k surface destroys the model. The volume loss term (~18,700) is ~40,000x larger than the surface loss (~0.48), creating gradient norms of ~2,500 (vs ~5 for surface-only). The model diverged to NaN after ep169, then continued training in NaN state through ep999.

**Run 2 vs Run 3 — Does volume context help at 16k surface?**
NO. Also catastrophically worse. 64.86% vs 14.31%. Even with matched 16k/16k budget, volume points are pure noise — they don't help and actively destabilize training through the same gradient explosion mechanism. Diverged to NaN after ep81.

**Run 3 vs Run 4 — 16k vs 50k surface-only throughput tradeoff?**
16k training gets more epochs (484 vs 46) but worse accuracy (14.31% vs 13.48%). Note: Run 4's 46 epochs is an artifact of full eval (no --max-eval-batches), which costs ~7.8 min/epoch vs ~0.74 min/epoch for Run 3 (with --max-eval-batches 200). The baseline champion got 467 epochs at 50k surface with --max-eval-batches 200. At equal epoch count, 50k surface >> 16k surface.

**Run 4 vs Baseline — Full eval throughput cost**
Full eval (no --max-eval-batches) yields only 46 epochs vs 467 epochs with --max-eval-batches 200. This 10x throughput penalty results in 13.48% vs 3.997%. Full eval is NOT viable for training; should only be used for final paper-facing checkpoint evaluation.

### Volume Training Failure Analysis

The root cause is **unweighted loss summation**. The code (train.py) computes:
```
total_loss = surface_loss + volume_loss
```
Volume MSE loss operates on raw flow field quantities (velocity, pressure, turbulent viscosity) across 16k volume points, yielding loss values of ~18,700. Surface loss on 50k points yields ~0.48. The volume term dominates by 40,000x, causing:
1. Gradient norms of ~2,200-2,500 (vs ~5 for surface-only)
2. The optimizer primarily optimizes volume predictions, ignoring surface
3. Eventually gradient explosion → NaN divergence (ep169 for Run 1, ep81 for Run 2)

**To make volume training viable, you would need:**
- Loss weighting (e.g. `total = surface_loss + 0.0001 * volume_loss`)
- Separate gradient clipping per group
- Potentially separate encoders (like AB-UPT's geometry-separated encoding)
- Volume loss normalization to match surface loss scale

### Commands Used

**Run 1:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --no-surface-only-drivaerml \
  --drivaerml-train-surface-points 50000 --drivaerml-train-volume-points 16000 \
  --drivaerml-eval-surface-points 50000 --drivaerml-eval-volume-points 16000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent emma --wandb-name "emma/dm-vol-run1-50ks-16kv" --wandb-group dm_volume_ablation
```

**Run 2:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --no-surface-only-drivaerml \
  --drivaerml-train-surface-points 16000 --drivaerml-train-volume-points 16000 \
  --drivaerml-eval-surface-points 50000 --drivaerml-eval-volume-points 16000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent emma --wandb-name "emma/dm-vol-run2-16ks-16kv" --wandb-group dm_volume_ablation
```

**Run 3:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 16000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --agent emma --wandb-name "emma/dm-vol-run3-16ks-only" --wandb-group dm_volume_ablation
```

**Run 4:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --agent emma --wandb-name "emma/dm-vol-run4-50ks-baseline-fulleval" --wandb-group dm_volume_ablation
```

### What happened — honest analysis

**Volume training is a clear negative result.** Both volume-augmented runs (Run 1, Run 2) diverged to NaN due to the massive loss scale mismatch between surface (~0.5) and volume (~18,700) losses. The gradient norms in volume runs (~2,200-2,500) are ~500x those in surface-only runs (~4-5). Without loss weighting or gradient clipping, the volume signal completely overwhelms surface learning.

The 50k surface-only baseline remains the best approach. The full-eval throughput cost (10x fewer epochs) confirms that --max-eval-batches 200 is essential for training and should only be removed for final checkpoint evaluation.

### Suggested follow-ups

1. **Volume loss weighting** — Re-run with `volume_loss_weight=0.001` or similar to match surface/volume loss scales. This is the most promising direction since AB-UPT likely uses weighted/normalized losses.
2. **Gradient clipping for volume runs** — Add `--grad-clip 1.0` to volume training runs. The champion DrivAerML config doesn't use grad-clip, but volume runs clearly need it.
3. **Separate volume head** — Train surface-only, then add a lightweight volume prediction head with frozen surface weights. Avoids gradient interference entirely.
4. **Volume loss normalization** — Normalize volume MSE by the variance of volume targets, bringing it to the same scale as surface loss.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3044 because ADVISOR: The 40,000x loss scale mismatch (volume MSE ~18,700 vs surface MSE ~0.48) makes unweighted volume training catastrophic on DM — grad norms 500x baseline. The diagnostic data is valuable (we now understand why volume needs explicit weighting), but the experimental direction is a dead end without volume_loss_weight=0.0001 or similar normalization. Closing this PR — properly weighted volume training would be a separate hypothesis..
